### PR TITLE
Add normalization of identifiers

### PR DESCRIPTION
**Pull request by ceddlyburge**
Replace repo with https://github.com/ceddlyburge/elm-representer/commit/3e29881a3584ab988730822b933f240ccde4da3b

This adds the normalization of identifiers, and keeps the normalization of whitespace and comments

We will probably remove the circleci config later, but I though it might be useful in the medium term while we decide what (if anything) to replace it with.

### CODE REVIEW

**Comment by iHiD** [comment]
Thanks for this. Is there a reason you've chosen to use Circle instead of GitHub Actions for the CI? It looks that that script could easily be ported to GHA and we use GHA everywhere now as our standard? Maybe @ErikSchierboom (or @cmccandless or @saschamann, who are good at these things) could do that for you if you don't want to? 🙂

File: bin/run.sh [hunk 1]
```diff
@@ -1,37 +1,33 @@
...
-WORK_DIR=/opt/representer/app
-cp -rp $INPUT_DIR $WORK_DIR
+# Normalize identifiers and remove comments
+ELM_FILES=("$INPUT_DIR"/*.elm)
+if [ ${#ELM_FILES[@]} == 0 ]; then
+    echo "No .elm file found (${ELM_FILES[@]})"
+    exit -1
+elif [ ${#ELM_FILES[@]} -gt 1 ]; then  
+    echo "Multiple .elm files found (${ELM_FILES[@]})"
+    exit -2
```
Comment (ErikSchierboom) [inline bin/run.sh:18]: This is absolutely fine. You could consider concatenating the representation of the multiple files, but there is little chance that their representation would ever match another solution using multiple files.

**Review by ErikSchierboom — APPROVED** [review]
Looks good! It is a bit too much code for me to properly review, but I saw lots of tests :)

**Comment by ceddlyburge** [comment]
> Thanks for this. Is there a reason you've chosen to use Circle instead of GitHub Actions for the CI? It looks that that script could easily be ported to GHA and we use GHA everywhere now as our standard? Maybe @ErikSchierboom (or @cmccandless or @SaschaMann, who are good at these things) could do that for you if you don't want to? 🙂

Hi Jeremy, the plan is to use GitHub actions before this request is merged, apologies for the confusion.

File: bin/run.sh (hunk 1 above)
Comment (ceddlyburge) [reply bin/run.sh:18]: Hi Erik, the code doesn't currently support multiple files. Concatenating the multiple normalizations of the code would be ok, but concatenating the mapping.json wouldn't work, as there could be duplicates. Potentially we should do something different than just returning an error though. Would it be better to just return the concatenated files as representation.txt and "@" for mapping.json?

Comment (mpizenberg) [reply bin/run.sh:18]: Hi, I still have to properly review the code but regarding multiple files, if returning an error is ok, I think we shouldn't care too much. @ceddlyburge , you could also use the previous version of this representer that strip comments, format the code and concatenate the normalized elm files.

Comment (ErikSchierboom) [reply bin/run.sh:18]: I think returning an error is fine.

**Comment by ceddlyburge** [comment]
> > Thanks for this. Is there a reason you've chosen to use Circle instead of GitHub Actions for the CI? It looks that that script could easily be ported to GHA and we use GHA everywhere now as our standard? Maybe @ErikSchierboom (or @cmccandless or @SaschaMann, who are good at these things) could do that for you if you don't want to? 🙂
> 
> Hi Jeremy, the plan is to use GitHub actions before this request is merged, apologies for the confusion.

Hi Jeremy, this is done now.

**Comment by iHiD** [comment]
Brilliant. Thanks! 🙂

**Comment by mpizenberg** [comment]
Sorry @ceddlyburge I've had a very busy pair of weeks and weekends. I hope I'll be able to review this in the coming week

**Comment by mpizenberg** [comment]
Hi Cedd, I've spent some time reviewing this PR and the representer seems to be working well! There are things that need changes though so I'll try to make those changes today and let you know when it's done so that you can review them.

## Regarding the representer

In the identifier mapping, no need to keep all things that do not change. So I've added and `initialIdentifierMapping` in the `State` to remove them at the end.
Also the mapping.json file written at the end was reversed of what the spec says, so I've reversed the keys and values for the final identifier mappings written at the end.

## Regarding the repo and build process

The `package.json` and `package-lock.json` are missing so the `npm install` command is failing.

In `make.sh` there is usage of `--debug` if the `-d` option was passed as argument. In fact we should not add that `--debug` flag since it's not needed to be able to call functions from the `Debug` module. It's only useful for web apps with an interface that want to use the "time travel debugger".

Also I think this would be clearer with a Makefile instead of that sh file, which I suppose comes from the template you used as a starting point. Maybe even better would be to add two separate node scripts to the package.json that we can run with `npm run ...`.

In `Dockerfile`, it is commented that `elm make` must be run previously in order to have a main.js file. I think this will cause problems with exercism automated build happening on AWS to enable the representer in the website server. So we'll have to make sure it's built in the Dockerfile. I don't think this is an issue since we have a multi-stage build process. Actually the comment that elm-format does not work on alpine is now obsolete for 0.8.5 I think. All in all I think we can improve that situation so I'll try rewriting the docker build process.

**Comment by mpizenberg** [comment]
I've updated the build scripts to use npm directly, now I'll have a look at the dockerfile.

**Review by mpizenberg — APPROVED** [review]
I've updated the docker container to build the elm code for the representer. And it's now based on node-alpine to shave some MBs (198Mb -> 132Mb)

I think all is good now.

**Comment by ceddlyburge** [comment]
Hi Matthieu

> The package.json and package-lock.json are missing so the npm install command is failing.

This was on purpose, as the github action uses elm-tooling, and it is recommended not to use these with the package.json (although it is still possible). It looks like this was more a problem with the readme being wrong, rather that the build being wrong. I think the package.json is only really being used to run one command, and possibly to install elm-tooling (although in the non recommended way). I wonder if this is worth it?

> In make.sh there is usage of --debug if the -d option was passed as argument. In fact we should not add that --debug flag since it's not needed to be able to call functions from the Debug module. It's only useful for web apps with an interface that want to use the "time travel debugger".

This was not my experience, I had to use the -d option to enable the Debug module, maybe we are using slightly different versions of Elm or something. If we work this out we should add it to the readme.

**Comment by mpizenberg** [comment]
Yeah regarding elm-tooling-action, I think I was a bit too cautious. After
discussing with Simon Lydell, there us no issue. The only potential problem
would be using a version of tool not supported yet by the action. Then I'd
need to update the action to use the latest version of elm-tooling.

Installing elm-tooling is required for the build step of the docker. Though
we could install it manually via npm directly in the build steps. Both are
probably fine, but at least The version in used is specified with the lock
file.

Regarding the -d option. It was indeed necessary in the old setup because
without it, the compiler option used was --optimize which prevent usage the
Debug module. What I meant is that it's enough to remove the --optimize, no
need to add the --debug (which is for the time travel debugger)

**Pull request merged**